pkg/registry/core/service/storage/storage.go

```diff
@@ -99,3 +125,102 @@ func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.Updat
 	// subresources should never allow create on update.
 	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation, false, options)
 }
+
+// defaults fields that were not previously set on read. becomes an
+// essential part of upgrading a service
+func (r *GenericREST) defaultServiceOnRead(obj runtime.Object) error {
+	service, ok := obj.(*api.Service)
+	if ok {
+		return r.defaultSvcOnRead(service)
+	}
+
+	serviceList, ok := obj.(*api.ServiceList)
+	if ok {
+		return r.defaultServiceList(serviceList)
+	}
+
+	// this was not an object we can default
+	return nil
+}
+
+// defaults a service list
+func (r *GenericREST) defaultServiceList(serviceList *api.ServiceList) error {
+	if serviceList == nil {
+		return nil
+	}
+
+	for i, svc := range serviceList.Items {
+		err := r.defaultSvcOnRead(&svc)
+		if err != nil {
+			return err
+		}
+		serviceList.Items[i] = svc
+	}
+
+	return nil
+}
+
+// defaults a single service
+func (r *GenericREST) defaultSvcOnRead(service *api.Service) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		return nil
+	}
+
+	if service == nil {
+		return nil
+	}
+
+	if len(service.Spec.IPFamilies) > 0 {
+		return nil // already defaulted
+	}
+
+	requireDualStack := api.RequireDualStack
+	singleStack := api.SingleStack
+
+	// headless services
+	if len(service.Spec.ClusterIPs) == 1 && service.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		service.Spec.IPFamilies = []api.IPFamily{*r.primaryIPFamily}
+
+		// headless+selectorless
+		// headless+selectorless takes both families
+		if len(service.Spec.Selector) == 0 {
+			service.Spec.IPFamilyPolicy = &requireDualStack
+			if *r.primaryIPFamily == api.IPv4Protocol {
+				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv6Protocol)
+			} else {
+				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv4Protocol)
+			}
+		} else {
+			// headless w/ selector
+			// follows cluster configuration
+			if r.secondaryFamily != nil { // dualstack cluster
+				service.Spec.IPFamilies = append(service.Spec.IPFamilies, *r.secondaryFamily)
+				service.Spec.IPFamilyPolicy = &requireDualStack
+			} else {
+				service.Spec.IPFamilyPolicy = &singleStack
+			}
+		}
+
+	} else {
+		// headful
+		// make sure a slice exists to receive the families
+		service.Spec.IPFamilies = make([]api.IPFamily, len(service.Spec.ClusterIPs), len(service.Spec.ClusterIPs))
```

Comment:
Wait, aren't we fixing up objects that predate the new fields? Won't `ClusterIPs` always be `nil` here? (Likewise above on the headless check I guess.)

Reply:
No. strategy is called prior and the field normalization kicks in.